File: src/services/providers/keychain.ts
import { invoke } from "@tauri-apps/api/core";

/** Store an API key in the OS keychain under the given alias. */
export async function setApiKey(alias: string, key: string): Promise<void> {
  await invoke("set_api_key", { alias, key });
}

/** Retrieve an API key from the OS keychain. Throws if not found. */
export async function getApiKey(alias: string): Promise<string> {
  return invoke<string>("get_api_key", { alias });
}

/** Remove an API key from the OS keychain. */
export async function deleteApiKey(alias: string): Promise<void> {
  await invoke("delete_api_key", { alias });
}

File: src-tauri/src/commands/keychain.rs
use crate::error::AppError;
use keyring::Entry;

pub const SERVICE_NAME: &str = "app.mindeck";

/// Store an API key in the OS keychain.
/// `alias` is the provider id, e.g. "deepseek" or "qwen".
#[tauri::command]
pub fn set_api_key(alias: String, key: String) -> Result<(), AppError> {
    let entry = Entry::new(SERVICE_NAME, &alias)
        .map_err(|e| AppError::Keychain(e.to_string()))?;
    entry
        .set_password(&key)
        .map_err(|e| AppError::Keychain(e.to_string()))
}

/// Retrieve an API key from the OS keychain.
#[tauri::command]
pub fn get_api_key(alias: String) -> Result<String, AppError> {
    let entry = Entry::new(SERVICE_NAME, &alias)
        .map_err(|e| AppError::Keychain(e.to_string()))?;
    entry
        .get_password()
        .map_err(|e| AppError::Keychain(e.to_string()))
}

/// Delete an API key from the OS keychain.
#[tauri::command]
pub fn delete_api_key(alias: String) -> Result<(), AppError> {
    let entry = Entry::new(SERVICE_NAME, &alias)
        .map_err(|e| AppError::Keychain(e.to_string()))?;
    entry
        .delete_credential()
        .map_err(|e| AppError::Keychain(e.to_string()))
}
